### ISSUE TITLE
On binaryproto mean files for data transformation, or why do we allow blobs with zero dimensions?

### ISSUE THREAD

**Issue by Noiredd** [BVLC/caffe#6087]
Suppose I am trying to train some network using the CLI, with LMDB data source and a simple transformation, like so:
```
layer {
  name: "data"
  type: "Data"
  top: "data"
  transform_param { mean_file: "/media/ssd/DB/mean.binaryproto "}
  data_param {
    source: "/media/ssd/DB"
    batch_size: 32
    backend: LMDB
  }
}
```
I generated the mean file manually, using Python, and saved it to `.binaryproto` this way:
```python
#"image" is a 3-dimensional image (C, H, W; same as images - let's say 1x128x128)
blob = caffe.io.array_to_blobproto(image)
with open('mean.binaryproto', 'wb') as bfile:
  bfile.write(blob.SerializeToString())
```
This results in `Check failed: datum_channels == data_mean_.channels() (1 vs. 128)` ([thrown here](https://github.com/BVLC/caffe/blob/99466224dac86ddb86296b1e727794fb836bd80f/src/caffe/data_transformer.cpp#L62)). See, I didn't know that the mean image has to be 4-dimensional. What happened here is the blob got loaded as 3D (CHW) and aligned with the 4D data blob (NCHW) so that N_data was matched against C_mean, and C_data (1) against H_mean (128), which caused the mismatch. We could probably fix this by checking what are the data and mean dimensions and expand the mean to match - or at least provide more informative logs (or even CHECKs).  
Upon deeper inspection, this is even more confusing: we do not really use this 0-th dimension for the transformation itself, it just needs to be there for the sake of [loading the blob](https://github.com/BVLC/caffe/blob/99466224dac86ddb86296b1e727794fb836bd80f/src/caffe/blob.cpp#L466). So if we loaded a 2x3x128x128 mean blob, the second "image" would never be read.

In one of the attempts to work around this, I used the legacy shape descriptors:
```
blob.channels = image.shape[0]
blob.height = image.shape[1]
blob.width = image.shape[2]
```
which didn't help much, as I got `Check failed: data_` in [blob.cpp](https://github.com/BVLC/caffe/blob/99466224dac86ddb86296b1e727794fb836bd80f/src/caffe/blob.cpp#L133). Turns out I forgot to (or: didn't know I have to) set `blob.num` as well - and the default value of zero [causes](https://github.com/BVLC/caffe/blob/99466224dac86ddb86296b1e727794fb836bd80f/src/caffe/blob.cpp#L474-L484) the `Reshape` to not allocate any memory. We could implement some safety checks, i.e. if a binary blob specifies only some of the deprecated dimensions, the others could be 1 instead of zero (e.g. by pre-filling the `shape` vector with `1`s and only change the values from existing fields (`has_num()` etc.)). It seems like a more logical choice of a default value, doesn't it?

Finally, we could avoid this special case in `Blob::Reshape` itself, [right here](https://github.com/BVLC/caffe/blob/99466224dac86ddb86296b1e727794fb836bd80f/src/caffe/blob.cpp#L32) - by doing `CHECK_GT` instead. But we also have a [test](https://github.com/BVLC/caffe/blob/99466224dac86ddb86296b1e727794fb836bd80f/src/caffe/test/test_blob.cpp#L54-L60) in place that explicitly makes sure we can create blobs with some dimensions equal to zero. Where is that useful?

To sum up:

- [ ] What would be the right thing to do about `DataTransformer` mean binaryproto? Some guidelines in documentation, or runtime checks with proper feedback?
- [ ] Is it desired behavior to default to zero shape when reshaping a blob using legacy dimensions?
- [ ] Why do we need to be able to create blobs with some dimensions equal to zero?
